five_tips_about_testable_golang_code.md

```diff
@@ -0,0 +1,171 @@
+| patch | Yes | No |
+
+![that is how good Python is.](https://imgs.xkcd.com/comics/python.png)
+
+Mock(testify) in golang is good. Just not that *magic* as Python's. When I first learn how to use it, the thought occurred to my mind most frequently is "What, I need to do this manually?" and "If I need to do this manually, what are YOU doing, mock?"
+
+~~" I am Groot." Mock replied.~~
+
+Mock in python can automatically do two things, one is **replace current real code to mock code**, another is **verify calls happen as expected**.
+
+Mock in go testify can do the second job(verify calls) but not first one(replace real code to mock). That part has to be done manually. And manually replace have more limitation than magic replace.
+
+We can talk about the limitations in next sections, for this section, just simply forget golang mock is not as powerful as python's mock, is a good start to write untestable code in go.
+
+## 2. Interface who? we just pass struct itself.
+TL;DR, `struct` cannot be mocked when passing as a parameter.
+
+Let's start with a golang example:
+```go
+func addAggregation(search *elastic.SearchService) *elastic.SearchService {
+    search.Aggregation(/* put some variable here*/)
+    search.Aggregation(/*put some other variable here*/)
+
+    return search
+}
+```
+
+I removed some information to make it more clear. This function takes an `elastic.SearchService`, do something to it and return `elastic.SearchService`.
+
+Test a function like that in Python is easy. Define `mock.Aggregation` somewhere, pass this mock in, assert it be called with given parameter, then assert mock been returned. done.
+
+But we cannot do the same thing in golang.
+
+Remember, golang is a strong type language. If a function takes `elastic.SearchService` as the parameter, you cannot pass anything else. The only possible way is to create a real `elastic.SearchService` object, and do some tweaking in `Aggregation` method. It is complex, and when you seeking the solution about "how to create a real elastic.SearchService object", it is easy to get lost in chasing infinite rabbits in infinite holes. You will forget what you want at first. ~~Just like a bad marriage.~~
+
+Good guys have a solution for this. Instead of a struct, passing an interface can solve it.
+Still same function:
+```go
+func addAggregation(search *SearchServiceInterface) *SearchServiceInterface {
+    search.Aggregation(/* put some variable here*/)
+    search.Aggregation(/*put some other variable here*/)
+
+    return search
+}
+
+type SearchServiceInterface interface {
+    Aggregation(/* exactly same variable here*/)
+}
+```
+
+It looks almost the same, the only difference is instead of `elastic.SearchService`, we passed in an `ElasticRepositoryInterface`. For this example, the compiler would not complain about the object passed in if it has `Aggregation(/* correct type variable here*/)` method. that means we can pass a mock like this into it for testing:
```

Comment:
You said `ElasticRepositoryInterface` but I think you mean `SearchServiceInterface` based on what you wrote above